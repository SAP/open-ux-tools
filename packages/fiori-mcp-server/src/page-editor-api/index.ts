export * from './api';
export * from './sapuxFtfsFileIO';
export {
    TreeNode,
    TreeNodeProperty,
    PageAnnotations,
    findNodeByAnnotationNodeId,
    PropertyPath,
    findByPath
} from './parser';
export * from './project';
