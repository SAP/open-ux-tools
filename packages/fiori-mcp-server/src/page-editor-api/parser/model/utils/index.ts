export * from './annotations';
export * from './i18n';
export * from './object';
export * from './sort';
export * from './utils';
