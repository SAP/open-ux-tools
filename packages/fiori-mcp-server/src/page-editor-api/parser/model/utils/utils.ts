import i18next from 'i18next';
import type { TranslationBundles, SupportedAggregationAction, PropertyPath, PropertyMessage } from '../types';
import type { JSONSchema4 } from 'json-schema';
import type { ObjectAggregation } from '../ObjectAggregation';
import { AggregationActions, ValidationState, PropertyMessageType } from '../types';
import { getI18nBundleName, getRelevantI18nBundle, resolveI18nValue } from './i18n';
import type { ProjectType } from '@sap-ux/project-access';

export interface AppTranslationBundleData {
    bundles: TranslationBundles;
    projectType?: ProjectType;
}

/**
 * Method returns application specific i18n bundle data.
 * @returns {AppTranslationBundleData} i18n bundle data.
 */
export const getAppI18nBundle = (): AppTranslationBundleData => {
    return {
        bundles: {
            annotation: {},
            app: {},
            service: {},
            ui5: {}
        }
    };
};

export const getTechnicalIdFromPath = (path: PropertyPath, firstOnly = false): string | undefined => {
    const key = path[path.length - 1];
    if (key) {
        let id = key.toString();
        const separator = id.includes(':::') ? ':::' : '::';
        const parts = key.toString().split(separator);
        id = parts[0];
        if (!firstOnly && parts.length > 1) {
            // remove first entry
            parts.splice(0, 1);
            id = parts.join('/');
        }
        if (id) {
            return id;
        }
    }
};

/**
 * Method disables deletion of last annotation node if there is any custom nodes.
 * There can be cases when custom nodes referenced to last annotation node and by deleting last annotation node, we can not provide correct order.
 * @param {ObjectAggregation} aggregation Container aggregation.
 * @param {string} text I18n key for disabled deletion explanation.
 */
export const ensureLastNodeIsUndeletable = (aggregation: ObjectAggregation, text: string): void => {
    // Find if there is very last node
    // Should be undefined if there is two annotation nodes
    let lastNativeNode: ObjectAggregation | undefined;
    let nativeNodes = 0;
    let customNodeExists = false;
    for (const key in aggregation.aggregations) {
        if (!aggregation.aggregations[key].custom) {
            lastNativeNode = nativeNodes > 0 ? undefined : aggregation.aggregations[key];
            nativeNodes++;
        } else {
            customNodeExists = true;
        }
    }
    // Check if signle and last annotation node exists with actions
    if (!customNodeExists || !lastNativeNode || !lastNativeNode.actions) {
        return;
    }
    // Find deletion action
    const index = lastNativeNode.actions.findIndex(
        (action: SupportedAggregationAction | AggregationActions): boolean =>
            action === AggregationActions.Delete ||
            (typeof action === 'object' && action.type === AggregationActions.Delete)
    );
    if (index !== -1) {
        let deleteAction = lastNativeNode.actions[index];
        if (typeof deleteAction !== 'object') {
            // Simple string - make object which allow to disable action
            deleteAction = {
                type: deleteAction
            };
        }
        // Disable actiondeletion
        deleteAction.disabled = true;
        deleteAction.title = i18next.t(text);
        lastNativeNode.actions[index] = deleteAction;
    }
};

/**
 * Function validates extension using validation parameter by setting warning message to aggregation.
 * @param {ObjectAggregation} aggregation aggregation for which to set warning message.
 * @param {boolean} isValid validation result.
 * @param {string} description description to show in case extension is invalid.
 * @param {string} property name of the property
 */
export const validateExtension = (
    aggregation: ObjectAggregation,
    isValid: boolean,
    description: string,
    property?: string
): void => {
    if (!isValid) {
        aggregation.state = ValidationState.Invalid;
        aggregation.messages = [
            {
                text: description,
                type: PropertyMessageType.Warning
            }
        ];
        aggregation.inactive = true;
        if (property && aggregation.properties[property]) {
            addValidationMessages(aggregation.properties[property], [{ text: description }]);
        }
    }
};

/**
 * Function validates macros extension key by checking aggregation value of it.
 * @param {ObjectAggregation} aggregation aggregation for which to set warning message.
 * @param {unknown} value value of custom extension aggregation.
 */
export const validateMacrosExtension = (aggregation: ObjectAggregation, value = aggregation.value): void => {
    const extensionValue = value && typeof value === 'object' ? value : {};
    const isKeyDefined = 'key' in extensionValue && !!extensionValue.key;
    // validate custom extension key
    validateExtension(aggregation, isKeyDefined, i18next.t('PAGE_EDITOR_CUSTOM_EXTENSION_NO_KEY'), 'key');
};

/**
 * Method which formats and returns display/description text for anchor selection option.
 * Method resolves i18n translation for text if translation exists.
 * @param {string} i18nKey I18n entry for option.
 * @param {ObjectAggregation} aggregation Object aggregation.
 * @param {string} text Anchor label without i18n resolution.
 * @param {string} [id] Anchor ID.
 * @returns {string} Formatted text for related anchor option.
 */
export const formatAnchorDescriptionText = (
    i18nKey: string,
    aggregation: ObjectAggregation,
    text: string,
    id?: string
): string => {
    const appI18n = getAppI18nBundle();
    const i18nBundle = getRelevantI18nBundle(aggregation, appI18n.bundles, appI18n.projectType);
    const label = resolveI18nValue(text, i18nBundle) || text;
    return id
        ? i18next.t(i18nKey, {
              id,
              label
          })
        : label;
};

/**
 * Method which receives 'oneOf' entries from schema and converts to dropdown option entries.
 * @param {string} i18nKey I18n entry for option.
 * @param {ObjectAggregation} aggregation Object aggregation.
 * @param {JSONSchema4[]} [oneOf] One of entries from schema.
 * @returns {UIComboBoxOption[]} Options for dropdown.
 */
export const oneOfToDropdownOptions = (
    i18nKey: string,
    aggregation: ObjectAggregation,
    oneOf: JSONSchema4[] = []
): { key: string; text: string }[] => {
    return oneOf
        .filter((entry) => !entry.hidden)
        .map((option, index) => {
            const text = option.description || option.const || index.toString();
            return {
                key: option.const || index.toString(),
                text: formatAnchorDescriptionText(i18nKey, aggregation, text, option.const)
            };
        });
};

/**
 * Method returns updated schema for anchor property.
 * Method hides enum entry which is associated with passed aggregation to avoid anchoring extension to itself.
 * @param {ObjectAggregation} aggregation Aggregation object.
 * @param {JSONSchema4} originalSchema Schema of anchor property.
 * @returns {JSONSchema4} Modified schema.
 */
export const updateAnchorSchema = (aggregation: ObjectAggregation, originalSchema: JSONSchema4): JSONSchema4 => {
    const anchor = aggregation.name === 'position' ? aggregation.parent?.name : aggregation.name;
    if (originalSchema.enum) {
        // Copy to avoid sharing same reference across other custom columns
        const schema = { ...originalSchema, enum: [...originalSchema.enum] };
        // Handle enum
        const index = schema.enum.findIndex((entry) => entry === anchor);
        if (index !== -1) {
            schema.enum.splice(index, 1);
        }
        return schema;
    } else if (originalSchema.oneOf) {
        const appI18n = getAppI18nBundle();
        // Copy to avoid sharing same reference across other custom columns
        const schema = { ...originalSchema, oneOf: [...originalSchema.oneOf] };
        // Handle oneOf
        for (let i = 0; i < schema.oneOf.length; i++) {
            const entry = schema.oneOf[i];
            entry.bundle = getI18nBundleName(entry.custom, true, !!entry.annotationNodeId, appI18n.projectType);
            // Hide anchors which refers to itself
            if (entry.const === anchor) {
                schema.oneOf[i] = {
                    ...entry,
                    hidden: true
                };
            }
        }
        return schema;
    }
    return originalSchema;
};

/**
 * Method adds validation message to property or node instance.
 * @param instance Property or node instance.
 * @param messages Array of validation messages.
 */
export function addValidationMessages(instance: { messages?: PropertyMessage[] }, messages: PropertyMessage[]): void {
    if (!instance.messages) {
        instance.messages = [];
    }
    instance.messages.push(...messages);
}
