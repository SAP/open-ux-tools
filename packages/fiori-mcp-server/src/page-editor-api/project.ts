import {
    FileName,
    getCapServiceName,
    getMinimumUI5Version,
    getWebappPath,
    readCapServiceMetadataEdmx
} from '@sap-ux/project-access';
import type { ApplicationAccess, Manifest } from '@sap-ux/project-access';
import type { FileData } from '@sap/ux-specification/dist/types/src';
import { existsSync } from 'node:fs';
import { readFile, readdir } from 'node:fs/promises';
import { join } from 'node:path';
import type { FlexChangeFile } from './types';

/**
 * Method returns main service of the application.
 *
 * @param appAccess Application access.
 * @returns main service name
 */
export function getMainService(appAccess: ApplicationAccess): string {
    const appId = appAccess.getAppId();
    const project = appAccess.project;
    let mainService: string | undefined;
    if (appId === undefined) {
        const appIds = Object.keys(project.apps);
        mainService = project.apps[appIds[0]].mainService;
    } else {
        const app = project.apps[appId];
        if (!app) {
            throw new Error('ERROR_INVALID_APP_ID');
        }
        mainService = app.mainService;
    }
    return mainService ?? 'mainService';
}

/**
 * Retrieves the service name for the application.
 *
 * @param appAccess - The application access object
 * @returns A promise that resolves to the service name
 */
export async function getServiceName(appAccess: ApplicationAccess): Promise<string> {
    let serviceName = getMainService(appAccess);
    if (appAccess.projectType === 'CAPJava' || appAccess.projectType === 'CAPNodejs') {
        // get CDS service name
        serviceName = await getCapServiceName(appAccess.project.root, appAccess.app?.services?.[serviceName].uri ?? '');
    }
    return serviceName;
}

/**
 * Reads annotation files for the application.
 *
 * @param appAccess - The application access object
 * @returns A promise that resolves to an array of FileData objects
 */
export async function readAnnotationFiles(appAccess: ApplicationAccess): Promise<FileData[]> {
    const annotationData: FileData[] = [];
    const mainServiceName = getMainService(appAccess);
    const mainService = appAccess.app?.services?.[mainServiceName];
    if (!mainService) {
        return [];
    }
    if (mainService.uri && (appAccess.projectType === 'CAPJava' || appAccess.projectType === 'CAPNodejs')) {
        const serviceUri = mainService?.uri ?? '';
        if (serviceUri) {
            const edmx = await readCapServiceMetadataEdmx(appAccess.root, serviceUri);
            annotationData.push({
                fileContent: edmx,
                dataSourceUri: serviceUri
            });
        }
    } else {
        if (mainService.local) {
            const serviceFile = await readFile(mainService.local);
            annotationData.push({
                dataSourceUri: mainService.local,
                fileContent: serviceFile.toString()
            });
        }
        const { annotations = [] } = mainService;
        for (const annotation of annotations) {
            if (annotation.local) {
                const annotationFile = await readFile(annotation.local);
                annotationData.push({
                    dataSourceUri: annotation.local,
                    fileContent: annotationFile.toString()
                });
            }
        }
    }
    return annotationData;
}

/**
 * Retrieves the manifest file for the application.
 *
 * @param appAccess - The application access object
 * @returns A promise that resolves to the Manifest object or undefined if not found
 */
export async function getManifest(appAccess: ApplicationAccess): Promise<Manifest | undefined> {
    const absoluteWebappPath = await getWebappPath(appAccess.app.appRoot);
    const manifest = join(absoluteWebappPath, FileName.Manifest);
    if (!existsSync(manifest)) {
        return undefined;
    }
    const file = await readFile(manifest);
    return JSON.parse(file.toString());
}

/**
 * Returns the project's ui5version.
 * It reads the vesion from minUI5Version in the manifest.json.
 * If the minUI5Version is not a number, then 'latest' is returned as default value in case of exceptions.
 *
 * @param appAccess - application access object.
 * @returns Resolved UI5 version for passed application.
 */
export async function getUI5Version(appAccess: ApplicationAccess): Promise<string> {
    const manifest = await getManifest(appAccess);
    let ui5Version = manifest ? getMinimumUI5Version(manifest) : undefined;

    if (ui5Version !== undefined && Number.isNaN(parseFloat(ui5Version))) {
        ui5Version = 'latest';
    }

    return ui5Version ?? 'latest';
}

/**
 * Reads all flex change files from the application's "changes" directory.
 *
 * @param appAccess - application access object.
 * @returns A promise that resolves to an array of flex change files.
 */
export async function readFlexChanges(appAccess: ApplicationAccess): Promise<FlexChangeFile[]> {
    const changes: FlexChangeFile[] = [];
    const changesFolderPath = appAccess.app.changes;
    if (changesFolderPath && existsSync(changesFolderPath)) {
        const files = await readdir(changesFolderPath);
        for (const file of files) {
            const change = await readFile(join(changesFolderPath, file), 'utf8');
            changes.push({
                physicalFileName: file,
                fileContent: change
            });
        }
    }
    return changes;
}
