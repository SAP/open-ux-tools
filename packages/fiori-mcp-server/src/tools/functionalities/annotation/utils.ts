import { ApplicationAccess } from "@sap-ux/project-access";
import {
    FileName,
    getCapServiceName,
    getMinimumUI5Version,
    getWebappPath,
    readCapServiceMetadataEdmx
} from '@sap-ux/project-access';

/**
 * Method returns main service of the application.
 *
 * @param project = project
 * @param appId - application id
 * @returns main service name
 */
export function getMainService(appAccess: ApplicationAccess): string {
    const appId = appAccess.getAppId();
    const project = appAccess.project;
    let mainService: string | undefined;
    if (appId === undefined) {
        const appIds = Object.keys(project.apps);
        mainService = project.apps[appIds[0]].mainService;
    } else {
        const app = project.apps[appId];
        if (!app) {
            throw new Error('ERROR_INVALID_APP_ID');
        }
        mainService = app.mainService;
    }
    return mainService ?? 'mainService';
}


export async function getServiceName(appAccess: ApplicationAccess): Promise<string> {
    let serviceName = getMainService(appAccess);
    if (appAccess.projectType === 'CAPJava' || appAccess.projectType === 'CAPNodejs') {
        // get CDS service name
        serviceName = await getCapServiceName(appAccess.project.root, appAccess.app?.services?.[serviceName].uri ?? '');
    }
    return serviceName;
}