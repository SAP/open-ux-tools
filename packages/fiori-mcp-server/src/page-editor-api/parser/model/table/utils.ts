import { ArtifactType } from '@sap/ux-specification/dist/types/src';
import type { Location } from '../types';
import type { ObjectAggregation } from '../ObjectAggregation';

/**
 * Method checks if passed location is associated with annotation source/artifact.
 * @param {Location} location Location object.
 * @returns True if passed location is associated with annotation source/artifact.
 */
const isAnnotationLocation = (location: Location): boolean => {
    return !location.type || location.type === ArtifactType.Annotation;
};

/**
 * Method updates table and table's child nodes annotation "locations" properties.
 * It is used to handle different schema variations:
 * 1. "AnnotationPath" can be placed in "Table" node;
 * 2. "AnnotationPath" can be missed in "Table" node, but placed in child node;
 * In case of 2nd scenario - we need to copy "locations" from child and apply to "Table" node.
 * @param {ObjectAggregation} child Table's child aggregation/node.
 */
export const updateTableChildNodeLocations = (child: ObjectAggregation): void => {
    const annotationLocations: Location[] = [];
    const locations = child.locations ? [...child.locations] : [];
    // Remove annotation locations from original aggregation
    for (let i = locations.length - 1; i >= 0; i--) {
        const location = locations[i];
        if (isAnnotationLocation(location)) {
            annotationLocations.push(location);
            locations.splice(i, 1);
        }
    }
    // Add annotation locations to parent aggregation
    if (child.parent && annotationLocations.length) {
        const parentLocations = child.parent.locations || [];
        // Overwrite annotation locations if exists
        child.parent.locations = parentLocations.filter((location: Location) => !isAnnotationLocation(location));
        child.parent.locations.unshift(...annotationLocations);
    }
    child.locations = locations;
};
