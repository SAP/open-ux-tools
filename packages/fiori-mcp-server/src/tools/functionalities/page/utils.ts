import type { PageDef } from './types';
import { FioriElementsVersion, PageTypeV4 } from '@sap/ux-specification/dist/types/src';

const newPagePrefixForType = new Map<PageTypeV4, string>([
    [PageTypeV4.ListReport, 'List'],
    [PageTypeV4.ObjectPage, PageTypeV4.ObjectPage],
    [PageTypeV4.CustomPage, 'Page'],
    [PageTypeV4.FPMCustomPage, 'Page'],
    [PageTypeV4.AnalyticalListPage, 'ALP']
]);

/**
 * Generates the id for a new V4 page.
 *
 * @param page - page attributes.
 * @param parentPage - parent page; to check if it is a root page.
 * @param navigationProperty - Navigation property (optional).
 * @returns parts = an array of strings that shall be joined to form the page ID.
 */
function generatePageIdV4(page: PageDef, parentPage: string | undefined, navigationProperty?: string): string[] {
    let name = '';
    let parts: string[] = [];
    // Populate with page entity or name
    if (page.pageType === PageTypeV4.CustomPage || page.pageType === PageTypeV4.FPMCustomPage) {
        const viewId = page.viewName ?? '';
        const names = viewId.split('.');
        name = names[names.length - 1];
    } else {
        name = page.contextPath ? page.contextPath.slice(1).replace(/\//g, '_') : page.entitySet;
    }
    parts = [name];
    if (
        parentPage &&
        navigationProperty &&
        (!page.contextPath || page.contextPath.indexOf(navigationProperty) === -1)
    ) {
        parts.push(navigationProperty);
    }
    // Append page type to last part of page id - it places page type at the end of id without separator
    parts[parts.length - 1] += newPagePrefixForType.get(page.pageType as PageTypeV4);
    return parts;
}

/**
 * Generates the id for a new V2 page
 * @param {Page} page - page attributes
 * @returns parts = an array of strings that shall be joined to form the page ID
 */
function generatePageIdV2(page: PageDef): string[] {
    const parts: string[] = [];
    parts.push(page.pageType);
    if (page.entitySet) {
        parts.push(page.entitySet);
    }
    return parts;
}

/**
 * Generates the id for a new page.
 *
 * @param page - page attributes.
 * @param parentPage - parent page.
 * @param pages - Object with existing pages.
 * @param appVersion - fiori elements version.
 * @param navigationProperty - navigation property (optional).
 * @returns Generated page id.
 */
export function generatePageId(
    page: PageDef,
    parentPage: string | undefined,
    pages: PageDef[],
    appVersion: FioriElementsVersion,
    navigationProperty?: string
): string {
    let parts: string[] = [];
    if (appVersion === FioriElementsVersion.v2) {
        // Page generation for V2 page
        parts = generatePageIdV2(page);
    } else {
        // Page generation for V4 page
        parts = generatePageIdV4(page, parentPage, navigationProperty);
    }

    let pageId = parts.join('_');

    const existingPageIds = pages.map((existingPage) => existingPage.pageId);
    let idx = 0;
    while (existingPageIds.indexOf(pageId) > -1) {
        pageId += idx.toString();
        idx++;
    }
    return pageId;
}
