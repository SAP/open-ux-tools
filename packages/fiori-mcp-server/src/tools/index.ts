import type { Tool } from '@modelcontextprotocol/sdk/types.js';
import * as Input from '../types/input';
import * as Output from '../types/output';
import { convertToSchema } from './utils';

export { docSearch } from './hybrid-search';
export type { SearchResponseData } from './hybrid-search';
export { listFioriApps } from './list-fiori-apps';
export { listFunctionalities } from './list-functionalities';
export { getFunctionalityDetails } from './get-functionality-details';
export { executeFunctionality } from './execute-functionality';

export const tools = [
    {
        name: 'search_docs',
        title: 'Search in Fiori Documentation',
        description:
            "Searches code snippets of Fiori Elements, Annotations, SAPUI5, Fiori tools documentation for the given query. You MUST use this tool if you're unsure about Fiori APIs. Optionally returns only code blocks.",
        inputSchema: {
            type: 'object',
            properties: {
                query: {
                    type: 'string',
                    description: 'Search query'
                },
                maxResults: {
                    type: 'number',
                    description: 'Maximum number of results to return',
                    default: 25
                }
            },
            required: ['query']
        }
    },
    {
        name: 'list_fiori_apps',
        description: `Scans a specified directory to find existing SAP Fiori applications that can be modified.
                    This is an optional, preliminary tool.
                    **Use this first ONLY if the target application's name or path is not already known.**
                    The output can be used to ask the user for clarification before starting the main 3-step workflow.`,
        inputSchema: convertToSchema(Input.ListFioriAppsInputSchema),
        outputSchema: convertToSchema(Output.ListFioriAppsOutputSchema)
    },
    {
        name: 'list_functionality',
        description: `**(Step 1 of 3)**
                    Gets the complete and exclusive list of supported functionalities to create a new or modify an existing SAP Fiori application.
                    This is the **first mandatory step** to begin the workflow and requires a valid absolute path to a SAP Fiori application as input.
                    You MUST use a functionalityId from this tool's output to request details to the functionality in 'get_functionality_details' (Step 2).
                    You MUST not use a functionalityId as name of a tool.
                    Do not guess, assume, or use any functionality not present in this list, as it is invalid and will cause the operation to fail.
                    **Note: If the target application is not known, use the list_fiori_apps tool first to identify it.**
                    if the functionality list does not include a functionality to support the current goal, then try using the search_docs tool as a fallback.`,
        inputSchema: convertToSchema(Input.ListFunctionalitiesInputSchema),
        outputSchema: convertToSchema(Output.ListFunctionalitiesOutputSchema)
    },
    {
        name: 'get_functionality_details',
        description: `**(Step 2 of 3)**
                    Gets the required parameters and detailed information for a specific functionality to create a new or modify an existing SAP Fiori application.
                    You MUST provide a functionalityId obtained from 'list_functionality' (Step 1).
                    The output of this tool is required for the final step 'execute_functionality' (Step 3).`,
        inputSchema: convertToSchema(Input.GetFunctionalityDetailsInputSchema),
        outputSchema: convertToSchema(Output.GetFunctionalityDetailsOutputSchema)
    },
    {
        name: 'execute_functionality',
        description: `**(Step 3 of 3)**
                    Executes a specific functionality to create a new or modify an existing SAP Fiori application with provided parameters.
                    This is the **final step** of the workflow and performs the actual creation or modification.
                    You MUST provide the exact parameter information obtained from get_functionality_details (Step 2).`,
        inputSchema: convertToSchema(Input.ExecuteFunctionalityInputSchema),
        outputSchema: convertToSchema(Output.ExecuteFunctionalityOutputSchema)
    }
] as Tool[];
