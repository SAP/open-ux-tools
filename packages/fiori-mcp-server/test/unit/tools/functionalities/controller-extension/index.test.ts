import * as openUxProjectAccessDependency from '@sap-ux/project-access';
import {
    CREATE_CONTROLLER_EXTENSION_FUNCTIONALITY,
    createControllerExtensionHandlers
} from '../../../../../src/tools/functionalities/controller-extension';
import { mockSpecificationImport } from '../../../utils';
import * as toolUtils from '../../../../../src/tools/utils';
import * as projectUtils from '../../../../../src/page-editor-api/project';
import { join } from 'path';

jest.mock('@sap-ux/project-access', () => ({
    __esModule: true,
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
    ...(jest.requireActual('@sap-ux/project-access') as object)
}));

describe('create-controller-extension', () => {
    const appPath = join('root', 'app1');
    let importProjectMock = jest.fn();
    let generateCustomExtensionMock = jest.fn();
    const findProjectRootSpy: jest.SpyInstance = jest.spyOn(openUxProjectAccessDependency, 'findProjectRoot');
    const getDefaultExtensionFolderSpy: jest.SpyInstance = jest.spyOn(toolUtils, 'getDefaultExtensionFolder');
    const getUI5VersionSpy: jest.SpyInstance = jest.spyOn(projectUtils, 'getUI5Version');
    const getManifestSpy: jest.SpyInstance = jest.spyOn(projectUtils, 'getManifest');
    const createApplicationAccessSpy: jest.SpyInstance = jest.spyOn(
        openUxProjectAccessDependency,
        'createApplicationAccess'
    );
    const memFsdumpMock = jest.fn();
    beforeEach(async () => {
        importProjectMock = jest.fn().mockResolvedValue([]);
        memFsdumpMock.mockReturnValue({
            'manifest.json': {}
        });
        generateCustomExtensionMock = jest.fn().mockResolvedValue({
            commit: jest.fn(),
            dump: memFsdumpMock
        });
        getDefaultExtensionFolderSpy.mockReturnValue('controllerFolder');
        findProjectRootSpy.mockImplementation(async (path: string): Promise<string> => path);
        getUI5VersionSpy.mockResolvedValue('1.108.1');
        getManifestSpy.mockResolvedValue({});
        createApplicationAccessSpy.mockImplementation((rootPath: string) => {
            return {
                getAppId: () => 'app1',
                project: {
                    apps: {
                        ['app1']: {}
                    },
                    root: 'root'
                },
                getSpecification: () => ({
                    importProject: importProjectMock,
                    generateCustomExtension: generateCustomExtensionMock
                })
            };
        });
    });

    describe('getFunctionalityDetails', () => {
        test('getFunctionalityDetails', async () => {
            mockSpecificationImport(importProjectMock);
            const details = await createControllerExtensionHandlers.getFunctionalityDetails({
                appPath: appPath,
                functionalityId: CREATE_CONTROLLER_EXTENSION_FUNCTIONALITY.id
            });
            expect(details).toEqual({
                description:
                    'Add new controller extension by creating javascript or typescript file and updates manifest.json with entry. Controller extensions allow users to extensiate default behaviour with custom controllers code.',
                id: 'create-controller-extension',
                name: 'Add new controller extension by creating javascript or typescript file and updates manifest.json with entry',
                parameters: [
                    {
                        defaultValue: 'ListReport',
                        description: 'Type of page',
                        id: 'pageType',
                        options: ['ListReport', 'ObjectPage'],
                        required: true,
                        type: 'string'
                    },
                    {
                        description: 'Name of new controller extension file',
                        id: 'controllerName',
                        required: true,
                        type: 'string'
                    },
                    {
                        description:
                            'If controller extenison should be assigned for specific page, then pageId should be provided',
                        id: 'pageId',
                        options: ['TravelList', 'TravelObjectPage'],
                        type: 'string'
                    }
                ]
            });
        });

        test('getFunctionalityDetails - unresolvable project', async () => {
            createApplicationAccessSpy.mockImplementation(() => {
                throw new Error('Test error');
            });
            mockSpecificationImport(importProjectMock);
            const details = await createControllerExtensionHandlers.getFunctionalityDetails({
                appPath: appPath,
                functionalityId: CREATE_CONTROLLER_EXTENSION_FUNCTIONALITY.id
            });
            expect(details).toEqual({
                ...CREATE_CONTROLLER_EXTENSION_FUNCTIONALITY,
                parameters: [
                    CREATE_CONTROLLER_EXTENSION_FUNCTIONALITY.parameters[0],
                    CREATE_CONTROLLER_EXTENSION_FUNCTIONALITY.parameters[2]
                ]
            });
        });
    });

    describe('executeFunctionality', () => {
        test('create controller extension with pageType', async () => {
            mockSpecificationImport(importProjectMock);
            const details = await createControllerExtensionHandlers.executeFunctionality({
                appPath: appPath,
                functionalityId: CREATE_CONTROLLER_EXTENSION_FUNCTIONALITY.id,
                parameters: {
                    pageType: 'ListReport',
                    controllerName: 'Dummy'
                }
            });
            expect(generateCustomExtensionMock).toHaveBeenCalledTimes(1);
            expect(generateCustomExtensionMock).toHaveBeenCalledWith({
                basePath: appPath,
                customExtension: 'ControllerExtension',
                data: {
                    extension: {
                        pageId: undefined,
                        pageType: 'ListReport'
                    },
                    folder: 'controllerFolder',
                    minUI5Version: '1.108.1',
                    name: 'Dummy'
                }
            });
            expect(details).toEqual(
                expect.objectContaining({
                    appPath: appPath,
                    changes: ['manifest.json'],
                    functionalityId: 'create-controller-extension',
                    message: 'Controller extension is created',
                    parameters: {
                        pageType: 'ListReport',
                        controllerName: 'Dummy'
                    },
                    status: 'success'
                })
            );
        });

        test('create controller extension with pageId', async () => {
            mockSpecificationImport(importProjectMock);
            const details = await createControllerExtensionHandlers.executeFunctionality({
                appPath: appPath,
                functionalityId: CREATE_CONTROLLER_EXTENSION_FUNCTIONALITY.id,
                parameters: {
                    pageId: 'TravelObjectPage',
                    controllerName: 'Dummy'
                }
            });
            expect(generateCustomExtensionMock).toHaveBeenCalledTimes(1);
            expect(generateCustomExtensionMock).toHaveBeenCalledWith({
                basePath: appPath,
                customExtension: 'ControllerExtension',
                data: {
                    extension: {
                        pageId: 'TravelObjectPage',
                        pageType: 'ObjectPage'
                    },
                    folder: 'controllerFolder',
                    minUI5Version: '1.108.1',
                    name: 'Dummy'
                }
            });
            expect(details).toEqual(
                expect.objectContaining({
                    appPath: appPath,
                    changes: ['manifest.json'],
                    functionalityId: 'create-controller-extension',
                    message: 'Controller extension is created',
                    parameters: {
                        pageId: 'TravelObjectPage',
                        controllerName: 'Dummy'
                    },
                    status: 'success'
                })
            );
        });

        test('create controller extension - no changes', async () => {
            memFsdumpMock.mockReturnValue({});
            mockSpecificationImport(importProjectMock);
            const details = await createControllerExtensionHandlers.executeFunctionality({
                appPath: appPath,
                functionalityId: CREATE_CONTROLLER_EXTENSION_FUNCTIONALITY.id,
                parameters: {
                    pageId: 'TravelObjectPage',
                    controllerName: 'Dummy'
                }
            });
            expect(generateCustomExtensionMock).toHaveBeenCalledTimes(1);
            expect(details).toEqual(
                expect.objectContaining({
                    appPath: appPath,
                    changes: [],
                    functionalityId: 'create-controller-extension',
                    message: 'Controller extension is not created',
                    parameters: {
                        pageId: 'TravelObjectPage',
                        controllerName: 'Dummy'
                    },
                    status: 'skipped'
                })
            );
        });

        test('create controller extension - unresolvable project', async () => {
            createApplicationAccessSpy.mockImplementation(() => {
                throw new Error('Test error');
            });
            mockSpecificationImport(importProjectMock);
            const details = await createControllerExtensionHandlers.executeFunctionality({
                appPath: appPath,
                functionalityId: CREATE_CONTROLLER_EXTENSION_FUNCTIONALITY.id,
                parameters: {
                    pageId: 'TravelObjectPage',
                    controllerName: 'Dummy'
                }
            });
            expect(generateCustomExtensionMock).toHaveBeenCalledTimes(0);
            expect(details).toEqual(
                expect.objectContaining({
                    appPath: appPath,
                    changes: [],
                    functionalityId: 'create-controller-extension',
                    message: 'Controller extension is not created',
                    parameters: {
                        pageId: 'TravelObjectPage',
                        controllerName: 'Dummy'
                    },
                    status: 'skipped'
                })
            );
        });
    });
});
