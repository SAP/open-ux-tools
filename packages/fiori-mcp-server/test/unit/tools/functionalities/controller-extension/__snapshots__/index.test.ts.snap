// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`create-controller-extension getFunctionalityDetails getFunctionalityDetails 1`] = `
Object {
  "description": "Add new controller extension by creating javascript or typescript file and updates manifest.json with entry. Controller extensions allow users to extensiate default behaviour with custom controllers code.",
  "functionalityId": "create-controller-extension",
  "name": "Add new controller extension by creating javascript or typescript file and updates manifest.json with entry",
  "parameters": Object {
    "additionalProperties": false,
    "properties": Object {
      "controllerName": Object {
        "description": "Name of new controller extension file",
        "pattern": "^[A-Za-z][A-Za-z0-9_-]*$",
        "type": "string",
      },
      "pageId": Object {
        "description": "If controller extension should be assigned for specific page, then pageId should be provided",
        "enum": Array [
          "TravelList",
          "TravelObjectPage",
        ],
        "type": "string",
      },
      "pageType": Object {
        "default": "ListReport",
        "description": "Type of page",
        "enum": Array [
          "ListReport",
          "ObjectPage",
        ],
        "type": "string",
      },
    },
    "required": Array [
      "pageType",
      "controllerName",
    ],
    "type": "object",
  },
}
`;
